### DISCUSSION TITLE
컬렉션 프레임워크에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by kunsanglee** [2024-woowacourse-study/level-interview#62] [Java]
.

**Comment by kunsanglee** [comment]
컬렉션 프레임워크는 데이터 그룹을 효율적으로 처리하고 저장하기 위한 클래스와 인터페이스의 집합입니다.
1. Collection 인터페이스는 모든 컬렉션 클래스의 루트 인터페이스로, List, Set, Queue 등이 이 인터페이스를 상속합니다. 기본적인 컬렉션 작업을 정의합니다.
2. List 인터페이스는 순서가 있고, 중복을 허용합니다. ArrayList, LinkedList, Vector등이 있습니다.
3. Set 인터페이스는 중복을 허용하지 않으며, 순서를 보장하지 않습니다. HashSet, LinkedHashSet, TreeSet 등이 있습니다.
4. Queue 인터페이스는 요소를 처리하기 전에 보관하는 컬렉션을 나타냅니다. FIFO 또는 우선순위 큐 등의 특성을 가질 수 있습니다. LinkedList, PriorityQueue 등이 있습니다.
5. Map 인터페이스는 key-value 쌍으로 데이터를 저장합니다. 키는 중복이 없으며 각 키는 하나의 값과 매핑됩니다. HashMap, TreeMap, LinkedHashMap 등이 있습니다.

**Reply by kunsanglee** [reply]
멀티스레드 환경에서 발생할 수 있는 동시성 문제를 해결하기 위해 동기화된(Synchronized) 컬렉션을 제공합니다. 동기화된 컬렉션은 스레드 안전하게 설계되어 있어, 여러 스레드가 동시에 해당 컬렉션에 접근해도 데이터 일관성과 무결성을 유지할 수 있게 도와줍니다.

동기화를 지원하는 컬렉션 종류는 Vector와 Hashtable이 있습니다. Vector는 ArrayList의 동기화 버전이며, Hashtable은 HashMap의 동기화 버전입니다. 

Collections 유틸리티 클래스에는 비동기 컬렉션들을 동기화된 컬렉션으로 감싸는 정적 메서드를 제공합니다. synchronizedList, synchronizedMap, synchronizedSet이 있고, Collections.synchronizedList(new ArrayList<>()); 와 같은 형태로 동기화된 컬렉션을 얻을 수 있습니다. 

동기화된 컬렉션은 내부적으로 모든 메서드 호출을 synchronized 블록이나 메서드로 감싸서, 한 시점에 오직 하나의 스레드만 컬렉션을 수정할 수 있게 합니다. 이로써 동시성 문제를 해결할 수 있지만, 대규모 멀티스레드 환경에서 성능 저하의 원인이 되므로 잘 사용하지 않습니다.

**Reply by kunsanglee** [reply]
대규모 멀티스레드 환경에서의 성능이 떨어지는 동기화된 컬렉션의 대안으로 java 5부터 java.util.concurrent 패키지에 더 효율적인 동시성 컬렉션(Concurrent Collections)을 제공합니다.

동시성 컬렉션들은 내부적으로 데이터 구조를 세분화하여 여러 스레드가 동시에 접근할 때 전체에 대한 락이 아닌, 내부적으로 락 분할(lock splitting), 세거먼트 락(segmented locks), 비블로킹 알고리즘(non-blocking algorithms) 등의 기술을 사용하여 스레드 간의 경쟁을 최소화하여 성능을 향상시킵니다.

ConcurrentHashMap, ConcurrentLinkedQueue 등이 있습니다.

**Comment by seunghye218** [comment]
컬렉션은 객체 그룹을 나타내는 객체입니다. 컬렉션 프레임워크는 컬렉션을 표현하고 조작하기 위한 통합 아키텍처로, 구현 세부 사항과 별도로 컬렉션을 조작할 수 있습니다.

컬렉션 작업에 대한 표준화된 방식을 제공하므로 다양한 유형의 컬렉션에서 작업할 수 있는 코드를 더 쉽게 작성할 수 있습니다.

컬렉션 구현 클래스는 요소 추가, 제거, 액세스와 같은 일반적인 컬렉션 작업을 위한 효율적인 데이터 구조와 알고리즘을 제공합니다.

다양한 유형의 컬렉션에서 작동할 수 있는 코드를 작성할 수 있으므로 변화하는 요구 사항에 더 쉽게 적응할 수 있습니다.

요소 정렬, 검색 및 반복과 같이 컬렉션 작업을 더 쉽게 해주는 풍부한 메소드 및 유틸을 제공합니다.

**Reply by seunghye218** [reply]
컬렉션 프레임워크의 주요 장점은 다음과 같습니다.

1. 데이터 구조와 알고리즘을 제공하여 직접 작성할 필요가 없어 프로그래밍 노력을 줄여줍니다.
2. 데이터 구조 및 알고리즘의 고성능 구현을 제공하여 **성능을 향상시킵니다** . 각 인터페이스의 다양한 구현은 상호 교환이 가능하므로 구현을 전환하여 프로그램을 조정할 수 있습니다.
3. 컬렉션을 주고받을 수 있는 공통 언어를 설정하여 서로 관련이 없는 API 간의 상호 운용성을 제공합니다.
4. 여러 임시 컬렉션 API를 학습하도록 요구하여 **API 학습에 필요한 노력을 줄입니다 .**
5. 임시 컬렉션 API를 생성할 필요가 없으므로 **API를 설계하고 구현하는 데 필요한 노력이 줄어듭니다 .**
6. 컬렉션 및 컬렉션을 조작하는 알고리즘에 대한 표준 인터페이스를 제공하여 **소프트웨어 재사용을 촉진합니다 .**

https://docs.oracle.com/javase/8/docs/technotes/guides/collections/overview.html

**Reply by seunghye218** [reply]
컬렉션 프레임워크는 다음으로 구성됩니다. 

- Collection interfaces: 세트, 목록, 지도 등 다양한 유형의 컬렉션을 나타냅니다. 이러한 인터페이스는 프레임워크의 기초를 형성합니다.
- General-purpose implementations: 컬렉션 인터페이스의 기본 구현입니다.
- Legacy implementations: 이전 릴리스인 Vector 및 Hashtable 의 컬렉션 클래스는 컬렉션 인터페이스를 구현하도록 개조되었습니다.
- Special-purpose implementations: 특별한 상황에서 사용하도록 설계된 구현입니다. 이러한 구현은 비표준 성능 특성, 사용 제한 또는 동작을 표시합니다.
- Concurrent implementations: 고도의 동시 사용을 위해 설계된 구현입니다.
- Wrapper implementations: 다른 구현에 동기화와 같은 기능을 추가합니다.
- Convenience implementations: 컬렉션 인터페이스의 고성능 "미니 구현"입니다.
- Abstract implementations: 사용자 정의 구현을 용이하게 하기 위한 컬렉션 인터페이스의 부분 구현.
- Algorithms: 목록 정렬과 같은 컬렉션에 대한 유용한 기능을 수행하는 정적 메서드입니다.
- Infrastructure: 컬렉션 인터페이스에 대한 필수 지원을 제공하는 인터페이스입니다.
- Array Utilities: 기본 유형 및 참조 객체의 배열을 위한 유틸리티 함수입니다. 엄밀히 말하면 컬렉션 프레임워크의 일부는 아니지만 이 기능은 컬렉션 프레임워크와 동시에 Java 플랫폼에 추가되었으며 일부 동일한 인프라에 의존합니다.

https://docs.oracle.com/javase/8/docs/technotes/guides/collections/reference.html

**Comment by le2sky** [comment]
https://okky.kr/questions/1130916

**Comment by GIVEN53** [comment]
데이터 그룹
인터페이스
List, Set, Queue
Map -> Collection 구현 x
